### ISSUE TITLE
Feature Request: Installable package via winget

### ISSUE THREAD

**Issue by ngxson** [ggerganov/llama.cpp#8188]
### Prerequisites

- [X] I am running the latest code. Mention the version if possible as well.
- [X] I carefully followed the [README.md](https://github.com/ggerganov/llama.cpp/blob/master/README.md).
- [X] I searched using keywords relevant to my issue to make sure that I am creating a new issue that is not already open (or closed).
- [X] I reviewed the [Discussions](https://github.com/ggerganov/llama.cpp/discussions), and have a new and useful enhancement to share.

### Feature Description

On macos/linux, user can install a pre-built version llama.cpp easily via `brew`

It would be nice to have the equivalent to that on windows, via `winget`

### Motivation

The pre-built binary is already available via releases: https://github.com/ggerganov/llama.cpp/releases

It would be nice to somehow push them to https://winget.run/

However, I'm not familiar with working on windows, so I create this issue to further discuss and to look for help from the community.

**Comment by slaren** [comment]
The macOS situation is a bit easier to handle because there is only the Metal backend, and we know that it is available on every device. For windows and linux we would need different builds for CPU AVX, CPU AVX2, CPU AVX512, CUDA, HIP, SYCL, Vulkan,.. which is not very user friendly to say the least.

**Comment by gabrielgrant** [comment]
@slaren there are already builds for all those different compute acceleration options. Not sure I'm understanding why making the existing builds available via winget would make things more confusing?

**Comment by max-krasnyansky** [comment]
Ah. I didn't see this issue earlier. I kind of started looking into this already as well.
I wanted to publish winget packages for Windows on ARM64 (Snapdragon X-Elite) and a decently optimized x86-64 build.
For ARM64 we can just enable the CPU (ARMv8.7 NEON with MatMul INT8) for now. For x86-64 I was thinking of also publishing CPU with upto AVX512 for now.
Basically, at least enable folks to easily install a usable version, and they are looking for best perf they can install stuff from our releases directly. 
I don't know how well winget handles package flavors yet, perhaps there is a way to publish multiple packages for the same arch and have it pick one based on the machine details. 
If winget itself is not great at it we can add some wrapper scripts that fetch better versions.

**Comment by slaren** [comment]
I don't think that uploading a dozen different packages to winget is going to be a good experience for the users. It's barely tolerable as it is on the github releases, and here we assume that the users will be somehow technically competent enough to choose the right version. I would expect a winget package to be easier to use than what we can currently offer.

**Comment by gabrielgrant** [comment]
@slaren gotcha

@max-krasnyansky having a script to algorithmically determine which build is most appropriate would be great (whether for winget or even just for determining which of the builds on GH will run on a given machine)

**Comment by max-krasnyansky** [comment]
Sorry, if I wasn't clear.  My plan was to publish decent CPU versions to start so that simple `winget install llama.cpp` works.
Users get usable version with basically zero effort.

Then we can look into either having winget "auto-select" a better optimized version based on the user's machine
ie I'm definitely not suggesting user to do something like `winget install llama.cpp-cuda-v3.1-avx512` :)

If winget itself can't do that (I need to look a bit deeper into their package metadata format and options) then we can figure out something else. Maybe our own script, etc.

**Comment by AndreasKunar** [comment]
@max-krasnyansky : 
Ollama v0.3.12 supports winget install and it now also works great / native on my Snapdragon X Elite Surface Laptop 7 on Windows (for ARM). I did not look into the details, but it might be a good starting-point (since it builds on top of the llama.cpp base).

**Comment by max-krasnyansky** [comment]
> @max-krasnyansky : Ollama v0.3.12 supports winget install and it now also works great / native on my Snapdragon X Elite Surface Laptop 7 on Windows (for ARM). I did not look into the details, but it might be a good starting-point (since it builds on top of the llama.cpp base).

Yep. I saw that they have the `winget` package. I thought it was still x86-64 build though. Will take a look at the latest.

**Comment by slaren** [comment]
I think we are in a good position now to start releasing winget packages. I have been looking into how this could be done, and it looks fairly simple. Winget can automatically install a zip file to a private directory and add the executables to the path, so we don't really need do much beyond creating a manifest. An example of how to do this is the [cURL package](https://github.com/microsoft/winget-pkgs/tree/11f625cf914454c83adf117b0100d3c8b30373c6/manifests/c/cURL/cURL/8.9.1.3).

Once we create and submit the manifest, updates can be handled with a github action such as [this one](https://github.com/marketplace/actions/winget-releaser).

My plan is to start with a x64 package that includes only the Vulkan backend. This should be fairly safe since Vulkan is likely to work well enough on most systems. Currently including multiple GPU backends can cause issues when the same devices are supported by more than one backend, once this is resolved, the package will be extended to include other backends. Packages for Arm can also be added later.

**Comment by ngxson** [comment]
Btw no sure if this helps, but we can also allow the binary to download `.dll` using the curl functionality in `libcommon`

**Comment by slaren** [comment]
@ngxson This is handled by loading backends dynamically. After #13220 there will be a single release for the CPU build. We can't include every backend in a single package at the moment because of the problem I mentioned, but eventually that will be the goal.

**Comment by ericcurtin** [comment]
If winget does indeed use just Vulkan for now, it's a good start

**Comment by AndreasKunar** [comment]
> If winget does indeed use just Vulkan for now, it's a good start

Winget with a Vulkan backend might be a "universal" backend for Windows x64. For Windows arm64 I suggest to use only the CPU backend.

I think llama.cpp's advantage is its cutting-edge CPU and GPU optimizations - not sure if they are all possible just to be decided at runtime.

**Comment by slaren** [comment]
> For Windows arm64 I suggest to use only the CPU backend.

Would there be any advantage of also bundling the OpenCL backend in the arm64 package as well? Unless the model is offloaded with `-ngl`, the OpenCL backend wouldn't be active anyway, so it should be fairly safe to include it. Are there any Arm chips other than the Qualcomm Snapdragons X being used at the moment in Windows Arm64 devices?

**Comment by olegshulyakov** [comment]
There is already winget package with id `ggml.llamacpp`.

**Comment by rnett** [comment]
While the winget package looks nice, IMO it's a harmful footgun because it doesn't include CUDA (or any other specialized GPU backend AFAIK).  Despite this, and its performance implications, the docs and readme recommend it as the way to install on Windows.  However, doing so leads to worse performance than the various llama cpp wrappers (ollama, lm studio, etc) without the reason being obvious unless you notice in the logs that it's using Vulcan (and know what that means).

If you want to stay with winget, you could support using winget's `--custom` arg (which passes the args to the underlying app installer) to pass a list of backends or similar.  Alternatively, use `scoop` (or `chocolatey`) and their pre- and post- install script support.  Another alternative, although it's a bit of a hack, is to use another ecosystem's tools that supports feature selection, like https://github.com/cargo-bins/cargo-binstall (i.e. with an empty crate).

**Comment by gazzaf** [comment]
I was using the winget install for a week and began wondering why LLMs weren't as fast as expected, and then a particular model refused to run and it turns out it only works on CUDA so with that I discovered winget didn't install the version most appropriate for my hardware. It's my fault for assuming, but it would be great if it could support this --custom arg so CUDA or whatever can be specified. I appreciate the endless work that's gone into everything to do with llama.cpp.

**Comment by Supermanuu** [comment]
I don't know if this is an stupid comment but... What about using another ID for the CUDA backend? something like `ggml.llamacpp.cuda`?

**Comment by Alias4D** [comment]
when update dont get latest version 😒

wheb run : llama --version ---------------> b10217-ddd4ec142
where latest version now is b10545  ????????????
